File: Leetcode July Challenge/C++/Count Vowels Permutation.cpp
#include<unordered_map>
class Solution {
    unordered_map<int,vector<int>> m;
    int MOD = 1e9+7;
    

    public:

        int helper(int n, int i, char prev, vector<vector<long long> > &dp){

            if (i > n){
                return 0;
            }
            if(i == n){
                switch(prev)
                {
                    case 'a':
                        return 1;
                    case 'e':
                        return 2;
                    case 'i':
                        return 4;
                    case 'o':
                        return 2;
                    case 'u':
                        return 1;
                    default:
                        return 5;
                }
            }

            int idx = prev - 'a';
            if(dp[i][idx] != -1){
                return dp[i][idx];
            }
            long long ans = 0;
            for(auto next : m[prev]){
                ans += (helper(n, i+1, next, dp) % MOD);
            }


            return dp[i][idx] = ans % MOD;
        }

        int countVowelPermutation(int n) {
            m['c'] = {'a','e','i','o','u'};
            m['a'] = {'e'};
            m['e'] = {'a','i'};
            m['i'] = {'a','e','o','u'};
            m['o'] = {'i','u'};
            m['u'] = {'a'};

            vector<vector<long long> > dp(vector<vector<long long> >(n+2,vector<long long>(27,-1)));
            return helper(n , 1 , 'c', dp);

        }
};

File: Leetcode July Challenge/Python/Count Vowels Permutation.py
def countVowelPermutation(self, n: int) -> int:
        m = (10**9+7)
        from functools import lru_cache
        
        @lru_cache(None)
        def helper(last, n):
            if n==1:
                if last==None:
                    return 5
                else:
                    if last=='a':
                        return 1
                    if last=='e':
                        return 2
                    if last=='i':
                        return 4
                    if last=='o':
                        return 2
                    if last=='u':
                        return 1
                
            ans  = 0
            if last==None:
                ans += helper('a', n-1)+helper('e', n-1)+helper('i', n-1)+helper('o', n-1)+helper('u', n-1)
                ans %= m
            else:
                if last == 'a':
                    ans += helper('e', n-1)
                    ans %= m
                if last == 'e':
                    ans += helper('a', n-1)+helper('i', n-1)
                    ans %= m
                if last == 'i':
                    ans += helper('a', n-1)+helper('e', n-1)+helper('o', n-1)+helper('u', n-1)
                    ans %= m
                if last == 'o':
                    ans += helper('i', n-1)+helper('u', n-1)
                    ans %= m
                if last == 'u':
                    ans += helper('a', n-1)
                    ans %= m
                    
            return ans    
            
        return helper(None, n)
